syncd/syncd.cpp

```diff
@@ -3575,7 +3593,54 @@ int syncd_main(int argc, char **argv)
                  */
 
                 shutdownType = handleRestartQuery(*restartQuery);
-                break;
+                if (shutdownType != SYNCD_RESTART_TYPE_PRE_SHUTDOWN) {
+                    // break out the event handling loop to shutdown syncd
+                    break;
+                }
+
+                // Handle switch pre-shutdown and wait for the final shutdown
+                // event
+
+                SWSS_LOG_TIMER("warm pre-shutdown");
+
+                FlexCounter::removeAllCounters();
+                stopNotificationsProcessingThread();
+
+                sai_attribute_t attr;
+
+                attr.id = SAI_SWITCH_ATTR_RESTART_WARM;
```

Comment:
>SAI_SWITCH_ATTR_RESTART_WARM [](start = 26, length = 28)

What is current supporting status of different SAI implementations? If not supported, what will happen?

Reply:
Both 3.1 and 3.3 support SAI_SWITCH_ATTR_RESTART_WARM. 3.1 support is not officially complete.